src/robot/libdocpkg/model.py

```diff
@@ -190,3 +228,155 @@ def _arg_to_dict(self, arg):
+    def enum_catalog(self):
+        return sorted(self._enum_catalog)
+
+    @property
+    def typed_dict_catalog(self):
+        return sorted(self._typed_dict_catalog)
+
+    def add(self, typ):
+        type_doc = self._get_type_doc_object(typ)
+        if isinstance(type_doc, EnumDoc):
+            self._enum_catalog.add(type_doc)
+        elif isinstance(type_doc, TypedDictDoc):
+            self._typed_dict_catalog.add(type_doc)
+
+    def update(self, typs):
+        for typ in typs:
+            self.add(typ)
+
+    def _get_type_doc_object(self, typ):
+        if isinstance(typ, (EnumDoc, TypedDictDoc)):
+            return typ
+        if isinstance(typ, (TypedDictType, ExtTypedDictType)):
+            return TypedDictDoc.from_TypedDict(typ)
+        if isinstance(typ, EnumType):
+            return EnumDoc.from_Enum(typ)
+        if isinstance(typ, dict):
+            if typ.get('type', None) == 'TypedDict':
+                return TypedDictDoc.from_dict(typ)
+            if typ.get('type', None) == 'Enum':
+                return EnumDoc.from_dict(typ)
+        return None
+
+    def to_dictionary(self):
+        return {
+            'enums': [en.to_dictionary() for en in self.enum_catalog],
+            'typedDicts': [td.to_dictionary() for td in self.typed_dict_catalog]
+        }
+
+
+class TypedDictDoc(Sortable):
+
+    def __init__(self, name='', type='', doc='', items=None):
+        self.name = name
+        self.type = type
+        self.doc = doc
+        self.items = items or []
+
+    @classmethod
+    def from_dict(cls, type_doc):
+        if isinstance(type_doc, dict):
+            return cls(name=type_doc['name'],
+                       type=type_doc['type'],
+                       doc=type_doc['doc'],
+                       items=type_doc['items'])
+        raise TypeError(
+            'TypedDictDoc.from_dict() requires dictionary types but got %s.'
+            % type_name(type_doc))
+
+    @classmethod
+    def from_TypedDict(cls, typed_dict):
+        if isinstance(typed_dict, (TypedDictType, ExtTypedDictType)):
+            items = []
+            required_keys = list(getattr(typed_dict, '__required_keys__', []))
+            optional_keys = list(getattr(typed_dict, '__optional_keys__', []))
+            for key, value in typed_dict.__annotations__.items():
+                typ = value.__name__ if isclass(value) else unic(value)
+                items.append({'key': key,
+                              'type': typ,
+                              'required': key in required_keys
+                              if required_keys or optional_keys else None})
+            return cls(name=typed_dict.__name__,
+                       type='TypedDict',
+                       doc=typed_dict.__doc__ if typed_dict.__doc__ else '',
+                       items=items)
+        raise TypeError(
+            'TypedDictDoc.from_TypedDict() requires a TypedDict but got %s.'
+            % type_name(typed_dict))
```

Comment:
Are you sure this kind of error handling is needed? How likely is it that someone calls `from_TypedDict` or `from_dict` with an invalid argument? If such error handling really is needed, it should be done like

```python
if not valid:
    raise TypeError(...)
# several lines of code
```

not like
```python
if valid:
    # several lines of code
    return
raise TypeError(...)
```